pkg/vacuum/vacuum.go

```diff
@@ -196,70 +219,148 @@ func (e *Engine) Run(ctx context.Context) {
 			return
 		}
 		log.Info("msg", "compressed chunks need to be vacuumed", "count", len(chunks))
-		p := e.parallelism
-		if len(chunks) < p {
-			// don't spin up more workers than we could possibly use
-			// if parallelism is 6, but we only have 5 chunks to work on, use 5 workers
-			p = len(chunks)
+		numWorkers, err := e.calcNumWorkers(ctx, con, chunks)
+		if err != nil {
+			log.Error("msg", "failed to calc num workers", "error", err)
+			return
 		}
-		runWorkers(ctx, p, chunks, e.worker)
+		runWorkers(ctx, numWorkers, chunks, e.worker)
 		// in some cases, have seen it take up to 10 seconds for the stats to be updated post vacuum
 		time.Sleep(delay)
 	}
 }
 
 // listChunks identifies chunks which need to be vacuumed
-func (e *Engine) listChunks(ctx context.Context, con *pgxpool.Conn) ([]string, error) {
+func (e *Engine) listChunks(ctx context.Context, con *pgxpool.Conn) ([]*chunk, error) {
 	rows, err := con.Query(ctx, sqlListChunks)
 	if err != nil {
 		return nil, err
 	}
 	defer rows.Close()
-	tables := make([]string, 0)
+	chunks := make([]*chunk, 0)
 	for rows.Next() {
-		var table string
-		err := rows.Scan(&table)
+		c := chunk{}
+		err := rows.Scan(&c.name, &c.autovacuumCount, &c.age)
 		if err != nil {
 			return nil, err
 		}
-		tables = append(tables, table)
+		chunks = append(chunks, &c)
+	}
+	return chunks, nil
+}
+
+func (e *Engine) getVacuumSettings(ctx context.Context, con *pgxpool.Conn) (vacuumFreezeMinAge, autovacuumFreezeMaxAge int64, err error) {
+	err = con.QueryRow(ctx, sqlGetVacuumSettings).Scan(&vacuumFreezeMinAge, &autovacuumFreezeMaxAge)
+	if err != nil {
+		return 0, 0, err
+	}
+	return
+}
+
+func interpolate(vacuumFreezeMinAge, autovacuumFreezeMaxAge, maxChunkAge, maxParallelism float64) int {
```

Comment:
Worth adding some comment of what's the idea here

Reply:
https://github.com/timescale/promscale/pull/1761/commits/e06f3742ba4761e1c57f6778b7f2f3a9a0505c30